rlgym_tools/extra_state_setters/goalie_state.py

```diff
@@ -0,0 +1,164 @@
+        """
+        Function to set cars in preparation for an incoming shot
+
+        :param state_wrapper: StateWrapper object to be modified.
+        :param team_turn: team who's getting shot at
+        :param allow_enemy_interference: Boolean indicating whether opponents will spawn close enough to easily affect the play
+        :param first_defender_in_goal: Boolean indicating whether the first defender will spawn in the goal
+        :param reset_to_max_boost: Boolean indicating whether the cars will start each episode with 100 boost or keep from last episode
+        """
+        first_set = False
+        for car in state_wrapper.cars:
+            # set random position and rotation for all cars based on pre-determined ranges
+            
+            if car.team_num == team_turn and not first_set:   
+                if first_defender_in_goal:
+                    y_pos = -GOAL_LINE if car.team_num == 0 else GOAL_LINE
+                    car.set_pos(0, y_pos, z=17)
+                    first_set = True
+                else:
+                    self._place_car_in_box_area(car, team_turn)
+                    
+            else:
+                if allow_enemy_interference:
+                    self._place_car_in_box_area(car, team_turn)
+                
+                else:
+                    self._place_car_in_box_area(car, car.team_num)                    
+            
+            
+            car.set_rot(0, rand.random() * YAW_MAX - YAW_MAX/2, 0) 
+            
+            if reset_to_max_boost:
+                car.boost = 100
+            
+    def _place_car_in_box_area(self, car, team_delin):
+        """
+        Function to set a new ball towards a goal
+
+        :param car: car to be modified
+        :param team_delin: team number delinator to look at when deciding where to place the car
+        """
+    
+        y_pos = (PLACEMENT_BOX_Y - (rand.random() * PLACEMENT_BOX_Y)) 
+        
+        if team_delin == 0:
+            y_pos -= PLACEMENT_BOX_Y_OFFSET
+        else:
+            y_pos += PLACEMENT_BOX_Y_OFFSET
+    
+        car.set_pos(rand.random() * PLACEMENT_BOX_X - PLACEMENT_BOX_X/2, y_pos, z=17) 
+        
+    def _reset_ball(self, state_wrapper: StateWrapper, team_turn, aerial_only):
+        """
+        Function to set a new ball towards a goal
+
+        :param state_wrapper: StateWrapper object to be modified.
+        :param team_turn: team who's getting shot at
+        :param aerial_only: Boolean indicating whether should shots only be from the air
+        """
+        
+        pos, lin_vel, ang_vel = self._get_shot_parameters(team_turn, aerial_only)
+        state_wrapper.ball.set_pos(pos[0], pos[1], pos[2])
+        state_wrapper.ball.set_lin_vel(lin_vel[0], lin_vel[1], lin_vel[2])
+        state_wrapper.ball.set_ang_vel(ang_vel[0], ang_vel[1], ang_vel[2])
+        
+    def _get_shot_parameters(self, team_turn, aerial_only):
```

Comment:
this can be done more randomized in the future I guess. For now it is looking very nice